### PR TITLE
Paired Token Masking

### PR DESCRIPTION
**Pull request by ekiefl**
# Background

See https://www.notion.so/arcadiascience/Experimenting-with-notebook-publications-128c4f521e348096a490d4548a851818?pvs=4

# Check it out

Here is an HTML of the publication (unzip and open the root-level `index.html`): 
[html.zip](https://github.com/user-attachments/files/17497105/html.zip)

# The implementation

This repository corresponds to a prototype for the notebook publication format. If well received, we can create an Arcadia notebook publication template repo to streamline the process for future publications.

I'll describe a bit about how it works. Overall, the design philosophy is centered around making it *as easy as possible* to transform a Jupyter notebook into a publication. The scientist just writes a notebook in the repo root called `notebook.ipynb` whose contents will become the publication. When ready, they can generate a publication with `make pub`. The "publication" is a Sphinx docs webpage that in addition to the publication, contains a welcome page, license info, instructions for joining the analysis, and a code API where we can put additional code. Hypothes.is comments are enabled so we can receive inline feedback.

### CODE REVIEW

File: src/analysis/modal_esm/constants.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from __future__ import annotations
+
+import modal
+from rich.console import Console
+
+VOLUME: modal.Volume = modal.Volume.from_name("model-cache", create_if_missing=True)
+MODEL_STORAGE_PATH: str = "/vol"
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply src/analysis/modal_esm/constants.py:7]: Tbh I have no idea what role `/vol` plays but it is found in their examples/documentation so I'm pattern matching here.

I was surprised to find no mention of `/vol` when viewing the volume via their interface: https://modal.com/storage/arcadia-science/main/model-cache

File: src/analysis/modal_esm/cache_model.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+with IMAGE.imports():
+    from huggingface_hub import hf_hub_download
+    from huggingface_hub.utils import LocalEntryNotFoundError
+    from transformers import AutoModel, AutoTokenizer
+
+
+def model_exists(model_name: str) -> bool:
+    """Returns whether a HuggingFace model is cached in the remote volume.
+
+    The remote volume is hard-coded as `model-cache` and the storage path models is
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply src/analysis/modal_esm/cache_model.py:24]: Yep, that's my understanding of it. Here's some more details: https://modal.com/docs/reference/modal.Volume

File: src/analysis/modal_esm/cache_model.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        AutoTokenizer.from_pretrained(
+            model_name,
+            cache_dir=MODEL_STORAGE_PATH,
+            force_download=force_download,
+        )
+        CONSOLE.print(f"{model_name} tokenizer is now stored in volume.", style="green")
+        modified = True
+
+    if modified:
+        VOLUME.commit()
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply src/analysis/modal_esm/cache_model.py:112]: Not sure. Here is the source code:

```python
    @live_method
    async def commit(self):
        """Commit changes to the volume.

        If successful, the changes made are now persisted in durable storage and available to other containers accessing
        the volume.
        """
        async with self._lock:
            req = api_pb2.VolumeCommitRequest(volume_id=self.object_id)
            try:
                # TODO(gongy): only apply indefinite retries on 504 status.
                resp = await retry_transient_errors(self._client.stub.VolumeCommit, req, max_retries=90)
                if not resp.skip_reload:
                    # Reload changes on successful commit.
                    await self._do_reload(lock=False)
            except GRPCError as exc:
                raise RuntimeError(exc.message) if exc.status in (Status.FAILED_PRECONDITION, Status.NOT_FOUND) else exc
```

File: src/analysis/modal_esm/predict.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        attention_mask = attention_mask.repeat(batch_size, 1)
+
+        # Apply mask tokens in batch for each pair of positions
+        for batch_idx, mask_positions in enumerate(masked_positions):
+            for pos in mask_positions:
+                input_ids[batch_idx, pos] = self.tokenizer.mask_token_id
+
+        with torch.no_grad():
+            outputs = self.model(input_ids=input_ids, attention_mask=attention_mask)
+            return outputs.logits.cpu().numpy()
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply src/analysis/modal_esm/predict.py:141]: I'm pretty new to torch, so I'm not sure what best convention is (or what detach does). But after researching, it seems that detach is used to detach the tensor from the computation graph. But since this occurs within the `no_grad` context, there is no computation graph to detach from.

File: src/analysis/modal_esm/predict.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    """
+
+    download_and_cache_model.remote(model_name, force_download=False)
+
+    start_time = time.time()
+
+    CONSOLE.print(f"Using model {model_name}.")
+
+    predict_cls = ESMMaskedPrediction.with_options(gpu=gpu, concurrency_limit=num_gpus)  # type: ignore
+    worker_factory = predict_cls(model_name)
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply src/analysis/modal_esm/predict.py:201]: Yeah this sounds more... correct. It's not really a factory anyways, it is just capable of spawning GPU instances for prediction.

File: src/analysis/modal_esm/predict.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    CONSOLE,
+    HOURS,
+    IMAGE,
+    MINUTES,
+    MODEL_STORAGE_PATH,
+    PREDICTION_APP,
+    VOLUME,
+)
+
+with IMAGE.imports():
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply src/analysis/modal_esm/predict.py:27]: I agree that there are issues that arise because two environments must be managed. But if I understand you correctly, I think there might be a misconception. The value of this context manager is to hide packages from local that exist only on the remote (and thus raise ImportError locally), not to specify the complete set of packages modal will use. Modal code can reference packages imported outside of the managed import block.

There are different linting issues that arise because of this though. For example, we can't lint code for packages that don't exist locally. We also can't verify that the packages referenced by any modal-wrapped functions/classes actually exist in the image.

File: src/analysis/modal_esm/constants.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+IMAGE: modal.Image = (
+    modal.Image.debian_slim()
+    .apt_install("libpq-dev")
+    .pip_install("transformers", "torch", "pandas")
+)
+
+PREDICTION_APP: str = "esm-masked-prediction"
+CACHING_APP: str = "download-and-cache-models"
+
+CONSOLE: Console = Console()
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply src/analysis/modal_esm/constants.py]: Good call. I guess I don't have a clear cut view of when to use caps.

File: Makefile [hunk 8]
```diff
@@ -1,12 +1,44 @@
...
+.PHONY: test
+test:
+	pytest -v .
+
+# By default, `run-notebook` will just execute notebook.ipynb (overwriting
+# current state of notebook) using `jupyter nbconvert`. If the analysis
+# includes additional setup, `run-notebook` should be updated to include these
+# steps.
+.PHONY: run-notebook
+run-notebook:
```
[the remark this answers is not in the document]
Comment (ekiefl) [reply Makefile:26]: Yeah this is a fair question.

I reworked things in 10b731f753214e7eb7ce56c1a8ae91c905030d90 so that `make run-notebook` is externally facing. I did this by eliminating the need to deploy a permanent modal app as a setup step. Instead, ephemeral apps are created as needed. This is better anyways because you can treat your local as the head node instead of spawning a remote head node.

I don't want to force a potentially expensive re-run every time `make pub` is called, but I've better documented `make run-notebook` and why you want to use it.

File: src/analysis/modal_esm/constants.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (keithchev) [reply src/analysis/modal_esm/constants.py]: Hmm. Sounds like it's a de facto modal-specified constant, then. Kinda odd that they don't hide it. In any case, probs worth documenting here that it's not arbitrary and was copied from the modal docs.

File: src/analysis/modal_esm/constants.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (keithchev) [reply src/analysis/modal_esm/constants.py]: FWIW, my vibe on this is that truly immutable objects should be in all caps and objects that have state, manage state, or are interfaces to a service (like `app` or `volume` here, or a database `engine` in a web app) are generally in lower case.

[the remark this answers is not in the document]
Comment (keithchev) [reply src/analysis/modal_esm/constants.py]: Definitely a bit of a gray area, tho.

File: src/analysis/modal_esm/cache_model.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithchev) [reply src/analysis/modal_esm/cache_model.py]: Hmm. Kinda hard to tell, no? 🤔

File: src/analysis/modal_esm/predict.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (keithchev) [reply src/analysis/modal_esm/predict.py:141]: Ah, right. TIL! (I somehow thought there was always a graph...)

File: Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (keithchev) [reply Makefile:26]: Nice!

File: src/analysis/modal_esm/predict.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (keithchev) [reply src/analysis/modal_esm/predict.py]: Ah sorry yeah that wasn't clear. What I meant was basically your second point, that there's no way to check that there are no missing imports in the functions that will run on modal. Presumably, you'd have to take the union of the actual local imports, the imports specified in this context manager, and the imports within `app.function`-wrapped functions.

**Comment by ekiefl** [comment]
I've said it offline already, but it warrants saying it again: thanks again for this extensive review @keithchev. I feel that it exists in two separable parts: the **infrastructure**, and the **notebook content**.

**Infrastructure**: I've addressed all your comments, which I think provided some much needed polish that will serve us well if/when this repo gets templated.

**Notebook content**: There's some great ideas in here. Lots of your comments are easy to address and should be/will be implemented. Other comments request a great deal of work. As we've all discussed, there is an intrinsic tension between between agility and quality, and finding that balance is critical. In my opinion, your review exemplifies a level of scrutiny that dominantly prefers quality at the cost of agility. This may be a slight divergence from the goals intended from this publication format--at least my conception of them.

**Comment by ryanayork** [comment]
Great work @ekiefl! Also, I really appreciated reading your thorough review @keithchev. You bring up some important questions about the inherent tensions of this pub type. I think we can sort some of these tensions out now; others, I'm sure, will be ongoing areas of "creative" push and pull. My two cents is below, FWIW.

At a high level, these notebook pubs should lower the bar/speed up the release/expand the reach of our computational work. The goal is to make it easy for scientists to experiment with code and release their results _as they go_, allowing our work to get out faster _and_ letting the community weigh in earlier (including, at times, catching our mistakes!).

This means a couple of things. For one, the balance between the amount of **internal** and **external** reviews will likely differ from "typical" releases. We'll also need to embrace releasing a different stage of work. Specifically, to work as intended, these pubs will have less **internal** and (hopefully) more **external** review. They'll also be at an earlier/rougher/less confident stage. 

To me, a huge win would be having the ability to publish an idea quickly and get told how hopelessly wrong or led astray we are! If things are working, our research will constantly hit the ceiling of what we know @ Arcadia (and, best case scenario, what _anyone_ knows). If done right, these notebook pubs can be a vehicle for us to get rapid input on just how dumb or, again best case scenario, game-changing our work might be.

OK, (hopefully) these ideas sound nice. But how do we implement them? Some core questions, with proposed solutions:

- _How should we approach writing notebook pubs?_ 
-Thanks to @ekiefl, I think this question is already on it's way to being solved (at the very least, for a good chunk of it). This pub sets a standard for pub length, complexity, tone, etc. It's set to garner useful feedback from the community and will contribute some interesting ideas! 

- _How should code review be done?_
-This is the million-dollar question. @ekiefl I think your framing of **infrastructural** vs. **content** related comments helps clarify things. I'd argue that our notebook review philosophy should be weighted toward the **infrastructural** side. Does the code run, is it legible, and can it be reproduced? Luckily, the architecture of notebooks inherently addresses some of these issues and, best case scenario, should make the **infrastructural** review relatively straightforward most of the time. **Content** inherently takes longer to review, is more open to interpretation, and, I think, is exactly what we want the community to weigh in on. Ultimately, when we compile notebooks into full pubs, their impact will hopefully be multiplied via multiple rounds of external feedback. Some initial suggestions: 1) Code review for notebook pubs should focus mostly on the **infrastructural** side. 2) We should establish some basic and clear criteria for **content** (e.g., is the text reasonably clear?; are any figures legible?)

- How do we know they're ready for release?
-I think once the **infrastructure** of a notebook pub is legible and there aren't any glaring **content** issues, they should be ready for release. We should, as much as possible, try to minimize the amount of time it takes the author and reviewers to the pub there. **E.g. it would be amazing if we could get to the point where we could get these pubs out, from idea to release, in <1 week.** This may be crazy but, who knows, it's worth setting lofty goals. There are multiple things that may help. As with publishing v2 generally, managers should be constantly involved in the **content** side of things, hopefully obviating the need for extensive engagement there in the code review. I also think developing out **templates** for these pubs can help lower the amount that needs to be reviewed.

Just some ideas, looking forward to discussing further!

(...and this is all to say I think we should consider this pub as being close to release)

**Comment by keithchev** [comment]
@ekiefl:
> In my opinion, your review exemplifies a level of scrutiny that dominantly prefers quality at the cost of agility. This may be a slight divergence from the goals intended from this publication format--at least my conception of them.

Yeah, where to draw the line on the quality-vs-agility question is something I was/am struggling with. My bias is to emphasize quality because my intuition is that the attention and credibility we attract will likely be in proportion to the quality of what we publish. But this is just an intuition and I totally get that this is an experiment, so I'm *perfectly* content if you decide that any or all of my content-related comments are beyond the scope of pre-publication review for this new format. And of course I can always simply transfer any such comments to the pub itself.

@ryanayork:
> Some initial suggestions: 1) Code review for notebook pubs should focus mostly on the infrastructural side. 2) We should establish some basic and clear criteria for content (e.g., is the text reasonably clear?; are any figures legible?)

Agreed! Some rough guidelines for content would be great and what you suggest sounds reasonable to me, fwiw. And as I mentioned above, reviewers are also free to comment directly on the pub itself if they have comments about content that are beyond the scope of the guidelines.

> I also think developing out templates for these pubs can help lower the amount that needs to be reviewed.

Evan and I chatted about this and we think we can probably merge the infra Evan added in this repo into our existing `python-analysis-template` repo (with clear instructions on how to remove it for non-notebook-pub applications). If that proves unwieldy, we can always create a new template repo. In either case, future review of infra-related code should hopefully be minimal.

**Comment by ekiefl** [comment]
<p>Good points. After coming back with fresh eyes, I've found a way to better introduce the main concepts and the notebook's motivation.</p><p> </p><blockquote>I might be missing s/t, but at the end of this cell, aren't the two comparisons (1) and (2) the same? I think&nbsp;<code>p(i|{i,j})</code>&nbsp;should be equal to&nbsp;<code>p(i|{j,i})</code>, which is just a different way of writing<code>p(j|{i,j})</code></blockquote><p> </p><p>With <code>p(i|{i,j})</code>, you're asking what the distribution for position i is when i and j are masked.</p><p>With <code>p(j|{i,j})</code> , you're asking what the distribution for position j is when i and j are masked. </p><p> </p><p>These are different. I think the math notation is clear, but if you have recommendations for tightening it up let me know.</p>

---

 <a href='https://app.reviewnb.com/Arcadia-Science/2024-paired-token-masking/pull/1/discussion/#comment-2448697936'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ekiefl** [comment]
<p>I added some details.</p>

---

 <a href='https://app.reviewnb.com/Arcadia-Science/2024-paired-token-masking/pull/1/discussion/#comment-2448697946'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ekiefl** [comment]
<p>Good catch</p>

---

 <a href='https://app.reviewnb.com/Arcadia-Science/2024-paired-token-masking/pull/1/discussion/#comment-2448697963'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ekiefl** [comment]
<p>Good call.</p>

---

 <a href='https://app.reviewnb.com/Arcadia-Science/2024-paired-token-masking/pull/1/discussion/#comment-2448697975'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ekiefl** [comment]
<p>I feel like at minimum we need:</p><p> </p><p><code></code><code>python</p><p>single_logits_path = Path("output/logits_single.npz")</p><p>double_logits_path = Path("output/logits_double.npz")</p><p> </p><p>single_logits = get_logits_if_exists(single_logits_path)</p><p>double_logits = get_logits_if_exists(double_logits_path)</p><p></code><code></code></p><p> </p><p>At that point all we're leaving out is a very small function, which helps reveals the type of <code>single_logits</code> and <code>double_logits.</code></p>

---

 <a href='https://app.reviewnb.com/Arcadia-Science/2024-paired-token-masking/pull/1/discussion/#comment-2448697985'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ekiefl** [comment]
<p>Good catch!! PS I just learned that perplexity is independent of base, provided that the entropy and exponentiation use the <em>same</em> base, as you point out. </p>

---

 <a href='https://app.reviewnb.com/Arcadia-Science/2024-paired-token-masking/pull/1/discussion/#comment-2448698036'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by ekiefl** [comment]
<p>Ok I'm thinking bigger picture. I think basically all the calculation, including the modal code, can just be <code>calculate_or_load_logits</code>. Thinking back to Prachee who told me everyone's first publication at Arcadia is twice as long as it needs to be.  </p>

---

 <a href='https://app.reviewnb.com/Arcadia-Science/2024-paired-token-masking/pull/1/discussion/#comment-2448697985'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Review by keithchev — APPROVED** [review]
We chatted offline and agreed that, because the comments on this PR are already so extensive, we'll merge this PR as-is, and the changes that address the comments in this PR will be added in a subsequent PR.
